felix/calc/encapsulation_resolver.go

```diff
@@ -0,0 +1,207 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package calc
+
+import (
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/felix/config"
+	"github.com/projectcalico/calico/felix/dispatcher"
+
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+
+	"github.com/projectcalico/calico/libcalico-go/lib/backend/api"
+	"github.com/projectcalico/calico/libcalico-go/lib/backend/encap"
+	"github.com/projectcalico/calico/libcalico-go/lib/backend/model"
+)
+
+// EncapsulationResolver is a Calculation Graph component that watches IP pool updates and
+// calculates if the IPIP or VXLAN encaps should be enabled or disabled. If there is
+// a change, Felix restarts by calling configChangedRestartCallback() to apply them.
+type EncapsulationResolver struct {
+	config                       *config.Config
+	encapCalc                    *EncapsulationCalculator
+	configChangedRestartCallback func()
+	inSync                       bool
+}
+
+func NewEncapsulationResolver(config *config.Config, configChangedRestartCallback func()) *EncapsulationResolver {
+	if configChangedRestartCallback == nil {
+		log.Panic("Starting EncapsulationResolver with nil callback func.")
+	}
+
+	return &EncapsulationResolver{
+		config:                       config,
+		encapCalc:                    NewEncapsulationCalculator(config, nil),
+		configChangedRestartCallback: configChangedRestartCallback,
+		inSync:                       false,
+	}
+}
+
+func (r *EncapsulationResolver) RegisterWith(dispatcher *dispatcher.Dispatcher) {
+	dispatcher.Register(model.IPPoolKey{}, r.OnPoolUpdate)
+	dispatcher.RegisterStatusHandler(r.OnStatusUpdate)
+}
+
+func (r *EncapsulationResolver) OnPoolUpdate(update api.Update) (filterOut bool) {
+	if update.Value == nil {
+		log.WithField("update", update).Debug("EncapsulationResolver: IPPool deletion")
+
+		k, ok := update.Key.(model.IPPoolKey)
+		if ok {
+			r.encapCalc.RemoveModelPool(k)
+		} else {
+			log.Infof("failed to convert %+v to model.IPPoolKey. Ignoring.", update.Key)
+		}
+	} else {
+		log.WithField("update", update).Debug("EncapsulationResolver: IPPool update")
+
+		pool, ok := update.Value.(*model.IPPool)
+		if ok {
+			r.encapCalc.UpdateModelPool(pool)
+		} else {
+			log.Infof("failed to convert %+v to *model.IPPool. Ignoring.", update.Value)
+		}
+	}
+
+	r.TriggerCalculation()
+
+	return
+}
+
+func (r *EncapsulationResolver) OnStatusUpdate(status api.SyncStatus) {
+	log.WithField("status", status).Debug("EncapsulationResolver: SyncStatus update")
+
+	if status == api.InSync {
+		r.inSync = true
+		r.TriggerCalculation()
+	}
+}
+
+func (r *EncapsulationResolver) TriggerCalculation() {
+	// Do nothing if EncapsulationResolver hasn't sync'ed all updates yet
+	if !r.inSync {
```

Comment:
very small nit: I'd put the comment inside of the if statement (since the comment applies to the logic within the if statement).

Also probably worth a debug log indicating why we're skipping

Reply:
I was on the fence thinking logging could be too much from the updates, but you're right, they would be valuable... will add